File: brady/two_sum.java
import java.util.Arrays;

// O(n) time, O(n) space
class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> pairs = new HashMap<>();

        for (int i = 0; i < nums.length; ++i) {
            
            // Check if the difference (u.e., target-nums[i]) has already been found, and if so get its index
            int difference = target - nums[i];
            if (pairs.containsKey(difference)) {
                return new int[]{ pairs.get(difference), i };
            }

            // Add the number and its index to the map
            pairs.put(nums[i], i);
        }
        return null;
    }
}

File: mohit/03/two-sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        if ( nums.size() == 2 ) {
            return {0,1};
        } else if ( nums.size() <= 500 ) {
            for ( int it = 0; it + 1 < nums.size(); it++ ) {
                for ( int jt = nums.size() - 1; jt > it; jt-- ) {
                    if ( nums[it] + nums[jt]  == target ) {
                        return {it, jt};
                    }
                }
            }
        } else {
            unordered_map<int,int> map;
            for(int i=0;i<nums.size();i++){
               if(map.count(target-nums[i])>0) return {map[target-nums[i]],i};
               map[nums[i]]=i; 
            }
        }
        return {};
    }
};
